### PR TITLE
crypto: fixes hang on boot due to lack of entropy

### PR DESCRIPTION
**Pull request by vivinperis**
fixes: #3124

### CODE REVIEW

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build?
(admin: you have the next options (make sure you checked the code):
"ok to test" to accept this pull request (and further changes) for testing
"test this please" for a one time test run
"add to whitelist" add author of a Pull Request to whitelist (globally, be careful, it means this user can trigger kira for any PR)
do nothing -> CI won't start)

File: lib/crypto.c [hunk 1]
```diff
@@ -31,6 +31,7 @@
 #include "thread-utils.h"
 #include "compat/openssl_support.h"
 
+#include <criterion/criterion.h>
```
Comment (bazsi) [inline lib/crypto.c:34]: Criterion is a test only library, don't use it here.

File: lib/crypto.c [hunk 2]
```diff
@@ -64,19 +65,18 @@ crypto_init(void)
...
-  if (RAND_status() < 0 || getenv("RANDFILE"))
+  if (getenv("RANDFILE"))
     {
       char rnd_file[256];
 
       RAND_file_name(rnd_file, sizeof(rnd_file));
       if (rnd_file[0])
         {
           RAND_load_file(rnd_file, -1);
+          RAND_write_file(rnd_file);
```
Comment (bazsi) [inline lib/crypto.c:76]: we don't need the write portion here, in the issue, what I meant on RAND_write_file() is that we do this at exit, see crypto_deinit() in the same file.

File: lib/crypto.c [hunk 3]
```diff
@@ -64,19 +65,18 @@ crypto_init(void)
...
           randfile_loaded = TRUE;
+          cr_assert(RAND_status() >= 0,
```
Comment (bazsi) [inline lib/crypto.c:78]: instead of cr_assert() use g_assert(). cr_assert is only for unit test programs.

File: lib/crypto.c [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include "thread-utils.h"
 #include "compat/openssl_support.h"
 
+
```
Comment (bazsi) [inline lib/crypto.c:34]: please remove this extra change.

File: lib/crypto.c [hunk 5]
```diff
@@ -73,10 +73,12 @@ crypto_init(void)
         {
           RAND_load_file(rnd_file, -1);
           randfile_loaded = TRUE;
+          if (RAND_status() < 0)
+            {
+              fprintf(stderr,
+                      "WARNING: a trusted random number source is not available, crypto operations will probably fail. Please set the RANDFILE environment variable.");
```
Comment (bazsi) [inline lib/crypto.c:79]: Ah, one last thing. We should add more details to this message, the user has already specified RANDFILE as this point, but it didn't provide enough entropy.

Can you please clarify this message a bit?

Thanks

**Comment by bazsi** [comment]
@alltilla can you reapprove this? I have requested a minor clarification of the error message text. thanks.

**Pull request merged**